common/models/User-Identity.js

```diff
@@ -40,96 +28,93 @@ export default function(UserIdent) {
-    userIdentityModel.findOne({
+    const filter = {
       where: {
         provider: getSocialProvider(provider),
         externalId: profile.id
       }
-    }, function(err, identity) {
-      if (err) {
-        return cb(err);
-      }
-      if (identity) {
-        identity.credentials = credentials;
-        return identity.updateAttributes({
-          profile: profile,
-          credentials: credentials,
-          modified: new Date()
-        }, function(err) {
-          if (err) {
-            return cb(err);
-          }
-          // Find the user for the given identity
-          return identity.user(function(err, user) {
-            // Create access token if the autoLogin flag is set to true
-            if (!err && user && autoLogin) {
-              return (options.createAccessToken || createAccessToken)(
-                user,
-                function(err, token) {
-                  cb(err, user, identity, token);
-                }
-              );
-            }
-            return cb(err, user, identity);
-          });
-        });
-      }
-      // Find the user model
-      var userModel = userIdentityModel.relations.user &&
-        userIdentityModel.relations.user.modelTo ||
-        loopback.getModelByType(loopback.User);
-
-      var userObj = options.profileToUser(provider, profile, options);
+    };
+    return userIdentityModel.findOne(filter)
+      .then(identity => {
+        // identity already exists
+        // find user and log them in
+        if (identity) {
+          identity.credentials = credentials;
+          const options = {
+            profile: profile,
+            credentials: credentials,
+            modified: new Date()
+          };
+          return identity.updateAttributes(options)
+            // grab user associated with identity
+            .then(() => identity.user())
+            .then(user => {
+              // Create access token for user
+              const options = {
+                created: new Date(),
+                ttl: user.constructor.settings.ttl
+              };
+              return user.accessTokens.create(options)
+                .then(token => ({ user, token }));
+            })
+            .then(({ token, user })=> {
+              cb(null, user, identity, token);
+            })
+            .catch(err => cb(err));
+        }
+        // Find the user model
+        const userModel = userIdentityModel.relations.user &&
+          userIdentityModel.relations.user.modelTo ||
+          loopback.getModelByType(loopback.User);
 
-      if (!userObj.email && !options.emailOptional) {
-        process.nextTick(function() {
-          return cb('email is missing from the user profile');
-        });
-      }
+        const userObj = options.profileToUser(provider, profile, options);
+        if (getSocialProvider(provider) !== 'github') {
+          return process.nextTick(() => cb(
+            new Error(
```

Comment:
Yes, I don't think there is a built in way to show a flash message through this API. This PR will rely on yours